sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go

```diff
@@ -532,12 +533,26 @@ func (em *ElementManager) StateForBundle(rb RunBundle) TentativeData {
 	return ret
 }
 
+// Residual represents the unprocessed portion of a single element to be rescheduled for processing later.
+type Residual struct {
+	Element []byte
+	Delay   time.Duration // The relative time delay.
+	Bounded bool          // Whether this element is finite or not.
+}
+
+// Residuals is used to specify process continuations within a bundle.
+type Residuals struct {
+	Data                 []Residual
+	TransformID, InputID string                // Prism only allows one SDF at the root of a bundledescriptor so there should only be one each.
+	MinOutputWatermarks  map[string]mtime.Time // Output watermarks (technically per Residual, but aggregated here until it makes a difference.)
```

Comment:
Does the string key reference the bundle's window? If so could we add this? Also, is the `Min` of its namesake related to that the minimum element timestamp is what we consider as an input or output Watermark? Finally, I struggled with `technically per Residual` and `aggregated`. Is there a nuance here that is important i.e. cases when it is not 1:1 per Residual. What is being aggregated?

Reply:
There's no such thing as a "per window watermark", and windows themselves are only meaningful at aggregation boundaries. As such, no, we won't add windows to this.

Per the model/protos: The MinOutputWatermarks are set on a per output PCollection basis based on the estimated watermarks provided by the user code. So that string key refers to to the local output identifier.

In practice, the watermark for a single transform is the same for all outputs.